### PR TITLE
Add log_model param to MLFlowlogger

### PR DESCRIPTION
**Pull request by AlessioQuercia**
## What does this PR do?
This PR adds a log_model parameter to MLFlowLogger as the one in WandbLogger (#9138).

The log_model parameter can be set to:
- "all": checkpoints are logged during training.
- True: checkpoints are logged at the end of training, except when 
`:paramref: '~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint.save_top_k' == "-1"` which also logs every checkpoint during training.
- False: no checkpoint is logged.

In particular, the function behavior is not exactly the same as in WandbLogger, since MLFlowClient().log_artifact takes different parameters as input:
- In the WandbLogger, the artifact is logged together with a metadata dictionary and some aliases.
- In the MLFlowLogger, it is not possible to log metadata and aliases together with the artifact, therefore (to keep consistency) a temporary folder is created, metadata.yml and aliases.txt are created inside of it, then the temporary folder is logged on mlflow together with the checkpoint artifact in the model/checkpoints/CHECKPOINT_NAME/ folder.

### Does your PR introduce any breaking changes? If yes, please list them.
None.

## Before submitting
- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Notes on tests:
When I run the tests locally with `make test`, the output is: 1 failed, 1860 passed, 400 skipped, 3 xfailed, 5859 warnings. The failed test is the following one:

> FAILED tests/plugins/environments/test_slurm_environment.py::test_default_attributes - Failed: DID NOT RAISE <class 'KeyError'>

However, when I run the tests locally with the same command as the one in the makefile (`python -m coverage run --source pytorch_lightning -m pytest pytorch_lightning tests pl_examples -v`), the output is: 1861 passed, 400 skipped, 3 xfailed, 5859 warnings. 

Moreover, if I run the test which fails with make test singularly with `python -m coverage run --source pytorch_lightning -m pytest tests/plugins/environments/test_slurm_environment.py -v `, the test passes.

I assume the error might be given by the fact that I am using Ubuntu on WSL2 on Windows, but I am not sure. Anyway, I think the tests should pass on remote.

## Did you have fun?
Coding this small PR was funny 😃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 
+- Added `log_model` parameter to `MLFlowLogger`, as the one in `WandbLogger` ([#9138](https://github.com/PyTorchLightning/pytorch-lightning/pull/9138))
```
Comment (tchaton) [inline CHANGELOG.md:13]:
```suggestion
- Added `log_model` parameter to `MLFlowLogger` ([#9138](https://github.com/PyTorchLightning/pytorch-lightning/pull/9138))
```

File: pytorch_lightning/loggers/mlflow.py [hunk 2]
```diff
@@ -266,3 +291,70 @@ def version(self) -> str:
...
+    def after_save_checkpoint(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> None:
+        # log checkpoints as artifacts
+        if self._log_model == "all" or self._log_model is True and checkpoint_callback.save_top_k == -1:
+            self._scan_and_log_checkpoints(checkpoint_callback)
+        elif self._log_model is True:
+            self._checkpoint_callback = checkpoint_callback
+
+    def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> None:
+        # get checkpoints to be saved with associated score
+        checkpoints = {
```
Comment (tchaton) [inline pytorch_lightning/loggers/mlflow.py:304]: Can this function can converted into a utility function and shared between loggers ?

File: pytorch_lightning/loggers/mlflow.py [hunk 3]
```diff
@@ -266,3 +291,70 @@ def version(self) -> str:
...
+            aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
+
+            # Artifact path on mlflow
+            artifact_path = f"model/checkpoints/{Path(p).stem}"
+
+            print(p)
+            print(artifact_path)
+
+            # Log the checkpoint
+            self.experiment.log_artifact(self._run_id, p, f"{artifact_path}")
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/mlflow.py:344]:
```suggestion
            self.experiment.log_artifact(self._run_id, p, artifact_path)
```
already a string

File: pytorch_lightning/loggers/mlflow.py [hunk 4]
```diff
@@ -266,3 +291,70 @@ def version(self) -> str:
...
+                },
+            }
+            aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
+
+            # Artifact path on mlflow
+            artifact_path = f"model/checkpoints/{Path(p).stem}"
+
+            print(p)
+            print(artifact_path)
+
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/mlflow.py:342]:
```suggestion
```

File: pytorch_lightning/loggers/mlflow.py [hunk 5]
```diff
@@ -266,3 +291,70 @@ def version(self) -> str:
...
+            # Log the checkpoint
+            self.experiment.log_artifact(self._run_id, p, f"{artifact_path}")
+
+            # Create a temporary directory to log on mlflow
+            with tempfile.TemporaryDirectory() as tmp_dir:
+                # Log the metadata
+                with open(f"{tmp_dir}/metadata.yml", "w") as tmp_file_metadata:
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/mlflow.py:349]:
```suggestion
                with open(f"{tmp_dir}/metadata.yaml", "w") as tmp_file_metadata:
```
just for consistency across codebase of the extension for yaml files

File: tests/loggers/test_mlflow.py [hunk 6]
```diff
@@ -181,6 +181,46 @@ def training_epoch_end(self, *args, **kwargs):
...
+@mock.patch("pytorch_lightning.loggers.mlflow.mlflow")
+@mock.patch("pytorch_lightning.loggers.mlflow.MlflowClient")
+def test_mlflow_log_model(client, mlflow, tmpdir):
+    """Test that the logger creates the folders and files in the right place."""
+
+    mlflow.run = None
+    model = BoringModel()
+
+    # test log_model=True
+    logger = MLFlowLogger(experiment_name="exp-id-1", run_name="run-name-1", log_model=True)
```
Comment (rohitgr7) [inline tests/loggers/test_mlflow.py:193]: can you parametrize the log_model and use if-else statements for required assertions?

File: pytorch_lightning/loggers/mlflow.py (hunk 2 above)
Comment (AlessioQuercia) [reply pytorch_lightning/loggers/mlflow.py:304]: Yes, I was wondering the same when writing it, but I was not sure where to put it. An idea is to split the function in:
<details>
  <summary markdown="span"

 `_scan_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> List[ModelCheckpoint]` 

</summary>

```
_scan_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> List[ModelCheckpoint]:
    # get checkpoints to be saved with associated score
    checkpoints = {
        checkpoint_callback.last_model_path: checkpoint_callback.current_score,
        checkpoint_callback.best_model_path: checkpoint_callback.best_model_score,
        **checkpoint_callback.best_k_models,
    }
    checkpoints = sorted((Path(p).stat().st_mtime, p, s) for p, s in checkpoints.items() if Path(p).is_file())
    checkpoints = [
        c for c in checkpoints if c[1] not in self._logged_model_time.keys() or self._logged_model_time[c[1]] < c[0]
    ]
    return checkpoints
```

</details>

which is generic and can be probably put inside `Base.py`, and:

<details>
  <summary markdown="span"

` _log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:`  for `MLFlowLogger.py`

 </summary>

```
_log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:
    # log iteratively all new checkpoints
    for t, p, s in checkpoints:
    metadata = {
        # Ensure .item() is called to store Tensor contents
        "score": s.item() if isinstance(s, torch.Tensor) else s,
        "original_filename": Path(p).name,
        "ModelCheckpoint": {
            k: getattr(checkpoint_callback, k)
            for k in [
                "monitor",
                "mode",
                "save_last",
                "save_top_k",
                "save_weights_only",
                "_every_n_train_steps",
                "_every_n_val_epochs",
            ]
            # ensure it does not break if `ModelCheckpoint` args change
            if hasattr(checkpoint_callback, k)
        },
    }
    aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
    
    # Artifact path on mlflow
    artifact_path = f"model/checkpoints/{Path(p).stem}"
    
    # Log the checkpoint
    self.experiment.log_artifact(self._run_id, p, f"{artifact_path}")
    
    # Create a temporary directory to log on mlflow
    with tempfile.TemporaryDirectory() as tmp_dir:
        # Log the metadata
        with open(f"{tmp_dir}/metadata.yml", "w") as tmp_file_metadata:
            yaml.dump(metadata, tmp_file_metadata, default_flow_style=False)
    
        # Log the aliases
        with open(f"{tmp_dir}/aliases.txt", "w") as tmp_file_aliases:
            tmp_file_aliases.write(str(aliases))
    
        # Log the metadata and aliases
        self.experiment.log_artifacts(self._run_id, tmp_dir, artifact_path)
    
    # remember logged models - timestamp needed in case filename didn't change (lastkckpt or custom name)
    self._logged_model_time[p] = t
```

</details>

<details>
  <summary markdown="span"

` _log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:`  for `WandbLogger.py`

 </summary>

```
_log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:
    # log iteratively all new checkpoints
    for t, p, s in checkpoints:
    metadata = (
        {
            "score": s,
            "original_filename": Path(p).name,
            "ModelCheckpoint": {
                k: getattr(checkpoint_callback, k)
                for k in [
                    "monitor",
                    "mode",
                    "save_last",
                    "save_top_k",
                    "save_weights_only",
                    "_every_n_train_steps",
                    "_every_n_val_epochs",
                ]
                # ensure it does not break if `ModelCheckpoint` args change
                if hasattr(checkpoint_callback, k)
            },
        }
        if _WANDB_GREATER_EQUAL_0_10_22
        else None
    )
    artifact = wandb.Artifact(name=f"model-{self.experiment.id}", type="model", metadata=metadata)
    artifact.add_file(p, name="model.ckpt")
    aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
    self.experiment.log_artifact(artifact, aliases=aliases)
    # remember logged models - timestamp needed in case filename didn't change (lastkckpt or custom name)
    self._logged_model_time[p] = t
```

</details>

File: pytorch_lightning/loggers/mlflow.py (hunk 4 above)
Comment (AlessioQuercia) [reply pytorch_lightning/loggers/mlflow.py:342]: Yes, I also removed those earlier locally, but I forgot to push the changes xD

File: tests/loggers/test_mlflow.py (hunk 6 above)
Comment (AlessioQuercia) [reply tests/loggers/test_mlflow.py:193]: I am not sure if I understood how, like the following?
<details>
  <summary markdown="span"

 `test_mlflow_log_model(client, mlflow, tmpdir, log_model):` 

</summary>

```
@mock.patch("pytorch_lightning.loggers.mlflow.mlflow")
@mock.patch("pytorch_lightning.loggers.mlflow.MlflowClient")
def test_mlflow_log_model(client, mlflow, tmpdir, log_model):
    """Test that the logger creates the folders and files in the right place."""

    model = BoringModel()

    client.return_value.log_artifact.reset_mock()
    client.return_value.log_artifacts.reset_mock()
    logger = MLFlowLogger(experiment_name="exp-id-1", run_name="run-name-1", log_model=log_model)
    trainer = Trainer(default_root_dir=tmpdir, logger=logger, max_epochs=2, limit_train_batches=3, limit_val_batches=3)
    trainer.fit(model)

    if log_model == True:
        # Checkpoint log
        client.return_value.log_artifact.assert_called_once()
        # Metadata and aliases log
        client.return_value.log_artifacts.assert_called_once()
    elif log_model == 'all':
            # Checkpoint log
            assert client.return_value.log_artifact.call_count == 2
            # Metadata and aliases log
            assert client.return_value.log_artifacts.call_count == 2
    elif log_model == False:
            # Checkpoint log
            assert not client.return_value.log_artifact.called
            # Metadata and aliases log
            assert not client.return_value.log_artifacts.called
```

</details>

I am not very familiar with tests, should I use MagicMock to instantiate log_model or should the function just be called 3 times (once per different value of `log_model`?

Comment (rohitgr7) [reply tests/loggers/test_mlflow.py:193]: yeah something similar. https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions

Comment (AlessioQuercia) [reply tests/loggers/test_mlflow.py]: Added this in the last commits, I modified it also for test_wandb.py

File: pytorch_lightning/loggers/mlflow.py (hunk 2 above)
Comment (AlessioQuercia) [reply pytorch_lightning/loggers/mlflow.py]: Added this in the last commits

File: pytorch_lightning/loggers/wandb.py [hunk 7]
```diff
@@ -262,18 +262,9 @@ def finalize(self, status: str) -> None:
         if self._checkpoint_callback:
             self._scan_and_log_checkpoints(self._checkpoint_callback)
 
-    def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> None:
```
Comment (awaelchli) [inline pytorch_lightning/loggers/wandb.py:265]: I propose to split this PR into to parts: 
1) make necessary refactors
2) add feature to mlflow

this makes sure we don't mix a feature for mlflow with changes to the wandb logger. we should be able to separate this cleanly.

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 
+- Added `log_model` parameter to `MLFlowLogger` ([#9138](https://github.com/PyTorchLightning/pytorch-lightning/pull/9138))
```
[the remark this answers is not in the document]
Comment (AlessioQuercia) [reply CHANGELOG.md:13]: Shouldn't the link point to the issue where this PR was discussed?

[the remark this answers is not in the document]
Comment (awaelchli) [reply CHANGELOG.md:13]: nope, it points to where the changes was made, which is this PR :)

[the remark this answers is not in the document]
Comment (awaelchli) [reply CHANGELOG.md:13]: because there can be several PRs for a single issue

File: pytorch_lightning/loggers/base.py [hunk 9]
```diff
@@ -349,6 +350,49 @@ def _add_prefix(self, metrics: Dict[str, float]):
 
         return metrics
 
+    def _scan_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> List[ModelCheckpoint]:
```
[the remark this answers is not in the document]
Comment (AlessioQuercia) [reply pytorch_lightning/loggers/base.py:353]: I totally agree on this, since the function is only used by two of the loggers. Where should I put it? In `base.py`, outside the `LightningLoggerBase` class?

What about the other two functions:
- `_scan_and_log_checkpoints(...)`
- `_log_checkpoints(...)`

They are also only used by those two loggers, should I also remove them from the base class?

File: pytorch_lightning/loggers/wandb.py (hunk 7 above)
Comment (AlessioQuercia) [reply pytorch_lightning/loggers/wandb.py:265]: Should I squash all the commits into these 2 commits, or use this PR for the refactors and open a new one for the mlflow feature?

Comment (awaelchli) [reply pytorch_lightning/loggers/wandb.py:265]: Good question. My suggestion is (how I would do it): 

1. mark this PR as draft for now (top right corner)
2. make a new branch with refactor only, open PR
3. once that PR is merged rebase this one on the changes and continue with the feature here

since we are not in a rush, we can do the refactor first and then the feature. 
Fyi in approx 3 weeks there will be a feature freeze before 1.5 gets released.

Comment (AlessioQuercia) [reply pytorch_lightning/loggers/wandb.py:265]: I created a refactoring PR #9312

**Comment by AlessioQuercia** [comment]
It is still work in progress, I am waiting for review to #9312 in order to proceed for this one.

**Comment by AlessioQuercia** [comment]
The base PR (https://github.com/Lightning-AI/lightning/pull/9312) for this PR has been merged into master, is this PR still needed?

**Comment by awaelchli** [comment]
@AlessioQuercia yes, I think continuing with #9138 is possible. We can reopen this if you would like to rebase it, or opening a fresh PR would also be fine.

**Comment by AlessioQuercia** [comment]
> @AlessioQuercia yes, I think continuing with #9138 is possible. We can reopen this if you would like to rebase it, or opening a fresh PR would also be fine.

I created a new PR (#15246) as I think I messed up when pulling the updated master into this branch.
